File: 2.  Problems Solutions/Reverse a Stack/Solution.cpp
//* https://practice.geeksforgeeks.org/problems/reverse-a-stack/1?utm_source=gfg&utm_medium=article&utm_campaign=bottom_sticky_on_article
//* https://www.geeksforgeeks.org/reverse-a-stack-using-recursion/

//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution
{
public:
    void pushAtBottom(stack<int> &s, int data)
    {
        if (s.empty())
        { // s.size() == 0
            s.push(data);
            return;
        }

        int top = s.top();
        s.pop();
        pushAtBottom(s, data);
        s.push(top);
    }
    void Reverse(stack<int> &s)
    {
        if (s.empty())
        { // s.size() == 0
            return;
        }

        int top = s.top();
        s.pop();
        Reverse(s);
        pushAtBottom(s, top);
    }
};

//{ Driver Code Starts.

int main()
{
    int T;
    cin >> T;
    while (T--)
    {
        int N;
        cin >> N;
        stack<int> St;
        for (int i = 0; i < N; i++)
        {
            int x;
            cin >> x;
            St.push(x);
        }
        Solution ob;
        ob.Reverse(St);
        vector<int> res;
        while (St.size())
        {
            res.push_back(St.top());
            St.pop();
        }
        for (int i = res.size() - 1; i >= 0; i--)
        {
            cout << res[i] << " ";
        }

        cout << endl;
    }
}
// } Driver Code Ends

File: 2.  Problems Solutions/Reverse a Stack/Solution.java
//* https://practice.geeksforgeeks.org/problems/reverse-a-stack/1?utm_source=gfg&utm_medium=article&utm_campaign=bottom_sticky_on_article
//* https://www.geeksforgeeks.org/reverse-a-stack-using-recursion/

//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        // taking testcases
        int t = Integer.parseInt(br.readLine());
        for (int i = 0; i < t; i++) {
            String str = br.readLine();

            // input n and d
            int n = Integer.parseInt(str);
            Stack<Integer> stack = new Stack<>();
            String str1 = br.readLine();
            String[] starr1 = str1.split(" ");
            // inserting elements in the array
            for (int j = 0; j < n; j++) {
                stack.push(Integer.parseInt(starr1[j]));
            }
            // calling reverse() function
            Solution.reverse(stack);
            for (int j : stack) {
                out.print(j + " ");
            }
            out.println();
        }
        out.close();
    }
}
// } Driver Code Ends

// User function Template for Java

class Solution {
    public static void pushAtBottom(Stack<Integer> s, int data) {
        if (s.isEmpty()) { // s.size() == 0
            s.push(data);
            return;
        }

        int top = s.pop();
        pushAtBottom(s, data);
        s.push(top);
    }

    static void reverse(Stack<Integer> s) {
        // add your code here
        if (s.isEmpty()) { // s.size() == 0
            return;
        }

        int top = s.pop();
        reverse(s);
        pushAtBottom(s, top);
    }
}